uniswap-v2/logics/impls/router/router.rs

```diff
@@ -0,0 +1,505 @@
+            .ok_or(RouterError::AddOverFlow2)?
+            .try_into().map_err(|_| RouterError::CastOverflow1)?;
+
+        Ok(amount_in)    
+    }
+
+    default fn _add_liquidity(
+        &self,
+        token_a: AccountId,
+        token_b: AccountId,
+        amount_a_desired: Balance,
+        amount_b_desired: Balance,
+        amount_a_min: Balance,
+        amount_b_min: Balance,
+    ) -> Result<(Balance, Balance), RouterError> {
+        if FactoryRef::get_pair(&self.data::<data::Data>().factory, token_a, token_b).is_none() {
+            FactoryRef::create_pair(&self.data::<data::Data>().factory, token_a, token_b)?;
+        };
+
+        let (reserve_a, reserve_b) = self._get_reserves(self.data::<data::Data>().factory, token_a, token_b)?;
+        if reserve_a == 0 && reserve_b == 0 {
+            return Ok((amount_a_desired, amount_b_desired))
+        }
+
+        let amount_b_optimal = self._quote(amount_a_desired, reserve_a, reserve_b)?;
+        if amount_b_optimal <= amount_b_desired {
+            if amount_b_optimal < amount_b_min {
+                return Err(RouterError::InsufficientBAmount)
+            }
+
+            Ok((amount_a_desired, amount_b_optimal))
+        } else {
+            let amount_a_optimal = self._quote(amount_b_desired, reserve_b, reserve_a)?;
+            assert!(amount_a_optimal <= amount_a_desired);
+            if amount_a_optimal < amount_a_min {
+                return Err(RouterError::InsufficientAAmount)
+            }
+
+            Ok((amount_a_optimal, amount_b_desired))
+        }
+    }
+
+    default fn _swap(
+        &mut self,
+        amounts: Vec<Balance>,
+        path: Vec<AccountId>,
+        to: AccountId,
+    ) -> Result<(), RouterError> {
+        let factory = self.data::<data::Data>().factory;
+
+        for i in 0..path.len()-1 {
+            let (input, output) = (path[i], path[i+1]);
+            let (token_0, _) = self._sort_tokens(input, output)?;
+            let amount_out: Balance = amounts[i+1];
+
+            let (amount_0_out, amount_1_out) = if input == token_0 {
+                (0, amount_out)
+            } else {
+                (amount_out, 0)
+            };
+
+            let to = if i < path.len() - 2 {
+                self._pair_for(factory, output, path[i+2])?
+            } else {
+                to
+            };
+
+            PairRef::swap(
+                &self._pair_for(factory, input, output)?,
+                amount_0_out,
+                amount_1_out,
+                to
+            )?;
+        }
+
+        Ok(())
+    }
+
+    default fn _pair_for(
+        &self,
+        factory: AccountId,
+        token_a: AccountId,
+        token_b: AccountId,
+    ) -> Result<AccountId, RouterError> {
+        let (token_0, token_1) = self._sort_tokens(token_a, token_b)?;
+
+        // Original Uniswap Library pairFor function calculate pair contract address without making cross contract calls.
```

Comment:
can you add this in top of the function as Rust doc instead of inside the code block ?